---
title: "3. Selecting Nanopore Adaptive Regions"
author: "Ellie Weise"
date: "2024-09-20"
output: html_document
---

## Overview
Once our epigenetic aging sites were selected, we generated a file for nanopore adaptive sequencing to target CpGs within the genome. Since Nanopore recommends targeting 1-10% of the genome, and our 87 CpG sites with a 1000 base buffer would amount to far less than that, we identified 247 'CpG rich' regions across the genome to sequence. These regions included all 87 of our targets along with ~15,000 other CpG sites identified across individuals in our whole-genome sequencing data. We also included a putative inversion region on Chromosome 12.

### Data/Library loading and basic dataset cleaning
```{r,warning=F}
#load libraries
library(quantsmooth)
library(tidyverse)
library(stringr)

#load data
df <- read.csv("Output/CpGList_LengthAge.csv")
ChromSwap <- read.csv("Input/ChromosomeSwap.csv")
load("Input/allCpGs_nosingletons.rda")
CpGlist_best_comb <- read.csv(file = "Output/CpGList_LengthAge.csv")

##Dataset cleaning ####
#remove CpGs with NAs for the model
nas <- colSums(is.na(allCpGs1))
allCpGs1 <- allCpGs1[,which(nas < 1)]
#write.csv(allCpGs1,"allCpGs_nonas.csv",append = F,quote = F)

#remove NW and mito data from data set
allCpGs1 <- allCpGs1[,which(!(grepl(pattern = "NW",
                                    colnames(allCpGs1),perl = T)))]
allCpGs1 <- allCpGs1[,which(!(grepl(pattern = "NC_009709.1",
                                    colnames(allCpGs1),perl = T)))]
```


### generating files for all CpGs and selected CpGs
```{r}
# making CpG file
CpGnames <- colnames(allCpGs1)
tmp <- data.frame(str_split_fixed(string = CpGnames,pattern = "\\.",n = 3))
CpGlocs <- tmp %>% 
  mutate(CHROM = paste(X1,X2,sep = "."),Type = "CpG") %>% 
  rename(MapInfo=X3) %>% 
  dplyr::select(CHROM,MapInfo,Type)

#alternative file with only best CpGs from model
CpGlocs <- df %>% 
  dplyr::select(CHROM,POS,ModelType) %>% 
  rename(MapInfo = POS,Type = ModelType)
```

### Identification of CpG rich regions in the genome (within 1000 bases, for our second run we adjusted this to 10,000)
```{r}
## ID CpG-rich regions (less than 1000b apart) ####
CpGRegion <- data.frame(matrix(nrow = 0, ncol = 3))
colnames(CpGRegion) <- c("CHROM","StartPOS","EndPOS")

## loop to select regions for each chromsome
for (k in 1:length(unique(CpGlocs$CHROM))) {
  # setting chromosomes and counting number of CpGs in that chromosome
  currentCHR <- unique(CpGlocs$CHROM)[k]
  CpG_Chr <- subset(CpGlocs,CpGlocs$CHROM == currentCHR)
  
  #set maximum distance between CpG sites for inclusion in the region
  maxdist <- 1000
  #need at least 2 CpGs in the chromosome for this to work (different loop for chromosomes with one site)
  if(length(CpG_Chr$CHROM) >= 2){
    # making sure position on chromosome (MapInfo) is an integer so the if statements work
    CpG_Chr$MapInfo <- as.integer(CpG_Chr$MapInfo)
    # arrange in order of location on chromosome
    CpG_Chr <- CpG_Chr %>% arrange(CHROM,MapInfo)
    
    #setting the first two CpGs in Chr to be the beginning of the first region
    #start 1
    mins <- c()
    mins <- CpG_Chr[1,2]
    #start 2
    maxes <- c()
    #if site 2 is further than distance from site 1, set it as the max for this CpG region
    if(CpG_Chr[2,2] - CpG_Chr[1,2] >= maxdist){
      maxes <- CpG_Chr[2,2]
    }
    #if it's within distance bases, set as temporary max and test the next CpG
    tmpmax <- CpG_Chr[2,2]
    
    if(length(CpG_Chr$CHROM) > 2){
      ## loop to look for islands
    for (i in 3:length(CpG_Chr$CHROM)) {
      #set current max
      #if previous set set a new maximum, this CpG is now a new miniumum
      if(CpG_Chr[i-1,2] - CpG_Chr[i-2,2] >= maxdist){
        mins <- c(mins,CpG_Chr[i,2])
        #tmpmax <- CpG_Chr[i,2]
      }
      
      #check if previous CpG distance is farther than max dist from this CpG
      #if yes this CpG is a new maximum
      if(CpG_Chr[i-1,2] - CpG_Chr[i-2,2] >= maxdist & CpG_Chr[i,2] - CpG_Chr[i-1,2] >= maxdist){
        maxes <- c(maxes,CpG_Chr[i,2])
      }
      
      # if it's not, then it's the new temporary maximum and test the next CpG
      if(CpG_Chr[i-1,2] - CpG_Chr[i-2,2] < maxdist){
        if(CpG_Chr[i,2] - CpG_Chr[i-1,2] < maxdist){
          tmpmax <- CpG_Chr[i,2]
        }
        
        # if CpG is within distance of the end of the chromosome, then we're at the end and need a final maximum
        if(CpG_Chr[i,2] - CpG_Chr[i-1,2] >= maxdist){
          maxes <- c(maxes,tmpmax)
          
        }
      }
    }
    
    # add in the chromosome ends if they are close enough to a CpG island
    if(CpG_Chr[length(CpG_Chr$CHROM),2] - CpG_Chr[length(CpG_Chr$CHROM)-1,2] >= maxdist){
      mins <- c(mins,CpG_Chr[length(CpG_Chr$CHROM),2])
    }
    
    maxes <- c(maxes,CpG_Chr[length(CpG_Chr$CHROM),2])
    }
    
    ## making data frame out of mins and maxes - these should be equal!!
    df <- data.frame(matrix(nrow = length(mins),ncol = 3))
    colnames(df) <- c("CHROM","StartPOS","EndPOS")
    
    df[,1] <- currentCHR
    df[,2] <- mins
    df[,3] <- maxes
    
    CpGRegion <- rbind(CpGRegion,df)
  }
  
  # make a data frame of all the chromosome and the CpG rich region min/max locations
  if(length(CpG_Chr$CHROM) == 1){
    df <- data.frame(matrix(nrow = 1,ncol = 3))
    colnames(df) <- c("CHROM","StartPOS","EndPOS")
    df[1,1] <- CpG_Chr$CHROM
    df[1,2] <- CpG_Chr$MapInfo
    df[1,3] <- CpG_Chr$MapInfo
    
    CpGRegion <- rbind(CpGRegion,df)
  }
}
```

### Adjusting the regions so they are at least 1000bp
*Nanopore sequencing requires a buffer region for adaptive sequencing otherwise not enough reads will be sequenced that contain the regions you're looking for*

```{r}
## Adjusting the regions so they cover at least 10000 bp
maxdist <- 1000
# loop by chromosome to adjust all regions to make sure they span at least the max dist size

for (i in 1:length(CpGRegion$CHROM)) {
  # calculate current size of the CpG rich region
  dist <- CpGRegion$EndPOS[i] - CpGRegion$StartPOS[i]
  
  # if it's less than maxdist, add an equal region on both sides so it reaches max dist
  if(dist < maxdist){
    # calculate how much to add to each side, round to integer
    addto <- round((maxdist - dist)/2,digits = 0)
    
    # make a new min and max
    newmin <- CpGRegion$StartPOS[i] - addto
    newmax <- CpGRegion$EndPOS[i] + addto
    # grab total length of the chromosome
    chrmax <- ChromSwap$Length[which(ChromSwap$NCseq == CpGRegion$CHROM[i])]
    
    # if we don't run into either chromosome end then we're good :)
    if(newmin >= 0 & newmax <= chrmax){
      CpGRegion$StartPOS[i] <- newmin
      CpGRegion$EndPOS[i] <- newmax
    }
    
    # if we hit the beginning of the chromosome, adjust to add more to the high end so the 
    # region is still maxdist length
    if(newmin < 0){
      newdist <- maxdist - CpGRegion$StartPOS[i] + 1
      newmax <- CpGRegion$EndPOS[i] + newdist
      CpGRegion$StartPOS[i] <- 1
      CpGRegion$EndPOS[i] <- newmax
    }
    
    # if we hit the end of the chromosome, adjust to add more to the low end so the region
    # is still maxdist length
    if(newmax > chrmax){
      newdist <- maxdist - (chrmax - CpGRegion$EndPOS)
      CpGRegion$StartPOS[i] <- CpGRegion$StartPOS - newdist
      CpGRegion$EndPOS[i] <- chrmax
    }
  }
}

#!# check if any of your adjusted regions now overlap!! Especially if your maxdist is very large
#!# more likely to happen the bigger the buffer region is
```


### Check nanopore size and amount of genome covered

```{r}
#size of all CpG regions to be amplified
CpGRegion$diff <- CpGRegion$EndPOS - CpGRegion$StartPOS

#comparing total size compared to full genome sequence (750MB)
sum(CpGRegion$diff)/750000000

## Removing sections related to inversion and then adding the full inversion region with a buffer
CpGregion_noinv <- CpGRegion[which(!(CpGRegion$CHROM == "NC_047158.1" & CpGRegion$StartPOS >= 4000000 & CpGRegion$EndPOS <= 13000000)),]

# inversion
colnames(CpGRegion)
inv <- c("NC_047158.1",as.integer(4000000),as.integer(13000000),as.integer(13000000-4000000))
length(CpGregion_noinv$CHROM)
#add a blank row
CpGregion_noinv[264,] <- c("NC_047158.1",as.integer(4000000),as.integer(13000000),as.integer(13000000-4000000))
CpGregion_addinv <- CpGregion_noinv %>% arrange(CHROM,as.integer(StartPOS))

sum(as.integer(CpGregion_addinv$diff))/750000000

#write CpG region table, should be formatted for use as a nanopore input file
#write.table(CpGregion_addinv[,1:3],file = "Output/NanoporeRegionSelection_040424.txt",quote = F,sep = "\t",row.names = F,col.names = F)

```

## Figures

### chromosomal visualization of CpG sites
```{r}
#make SNP panel plot####
CpGlocs$CHR<- NA
for (i in 1:length(ChromSwap$NCseq)) {
  CpGlocs$CHR[which(CpGlocs$CHROM == ChromSwap$NCseq[i])] <- ChromSwap$Einfeldt[i]
}

CpGlocs1 <- CpGlocs %>% dplyr::select(CHR,MapInfo,Type)
chrom <- ChromSwap[,3:4]
chrom <- chrom %>% rename(CHR = Einfeldt,MapInfo = Length) %>% mutate(Type = "ChrLength")


chrom <- rbind(chrom,CpGlocs1)
chrom$MapInfo <- as.integer(chrom$MapInfo)

## Making plot to show all CpGs and their location on the genome
tiff(filename = "Figures/CpGVisualization_032224.tiff",width = 84,height = 150,units = "mm",res = 400)
chrompos <- prepareGenomePlot(chrompos = chrom[,1:2], cols = "grey50",bleach = 0, topspace = 1, sexChromosomes = FALSE)
points(chrompos[25:length(chrom$CHR),2],chrompos[25:length(chrom$CHR),1]+0.05, , pch="|", cex = 0.75, col="deepskyblue4")
title("Atlantic Halibut CpG \nPositions by Chromosome")
dev.off()
```

